### ISSUE TITLE
caddy adapt fails to handle env address

### ISSUE THREAD

**Issue by ratorx** [caddyserver/caddy#4037]
When address is a runtime evaluated environment variable, `caddy adapt` fails with:
```
adapt: subject does not qualify for certificate: '{env.EXAMPLE}'
```

## Minimal Caddyfile
```
https://{env.EXAMPLE} {
	respond 200
}
```

## Expected behaviour
It does not fail :). It shouldn't use the raw placeholder value when adapt is called for the subject name check (that should be deferred to runtime, when the env is evaluated).

## Rough investigation
The error seems to be raised [here](https://github.com/caddyserver/caddy/blob/edb362aa96bd8e79adfaca44fbd9f9ce5bff778d/caddyconfig/httpcaddyfile/tlsapp.go#L205). I imagine that placeholders are handled generally by some other mechanism, but that mechanism isn't used here because the TLS app doesn't support placeholders (or expects them to have been evaluated before the config is passed into it).

**Comment by ratorx** [comment]
In case anyone else has the same problem, the workaround is to not use placeholders in the top level address and instead do something like:
```
https:// {
	@example {
		host {env.EXAMPLE}
	}
	handle @example {
		respond 200
	}
}
```

**Comment by francislavoie** [comment]
The `{env.*}` style of environment variables are replaced at runtime, which is too late to work as a hostname in Caddy. Instead, you may use the `{$ENV}` style which is a special placeholder in the Caddyfile which is replaced at config-adapt time, i.e. before Caddy starts running.

```
{$CADDY_HOST} {
	respond "It works!" 200
}
```

https://caddyserver.com/docs/caddyfile/concepts#environment-variables

For next time, please ask your usage questions on the [Caddy community forums](https://caddy.community). We prefer to keep the GitHub issue board for bugs and feature requests. Don't forget to fill out the thread template so we can help you!

**Comment by ratorx** [comment]
Sorry, I guess I could have been more clear. Why can hostnames not be resolved at runtime? Is runtime resolution a feature request that this project would consider, or is it technically infeasible?

I used GitHub because I felt it was either a feature request (could this work in future?) or an issue and it wasn't immediately clear from the docs that this was WAI.

**Comment by francislavoie** [comment]
The reason is explained right in the code you linked, it's necessary for the Caddyfile adapter to know the value at that point to know how to build the automation policies. It would be inconsistent if the Caddyfile adapter resolved `{env.*}` here but not elsewhere.

Why not just use the `{$ENV}` style? That's what it's there for.

**Comment by ratorx** [comment]
Unfortunately I can’t use `{$ENV}` style because I am using `caddy-l4`, so I convert my Caddyfile into JSON config (with `caddy adapt`) and merge it with the JSON config for `caddy-l4` and I don’t have access to the .env file I use at build time.

I’m currently not using the admin API to patch the config because that wouldn’t handle the recursive merging possible with `jq`. I’m also not generating the config at runtime, so I can validate it when creating my Docker image.

There are workarounds so that I could use the above, but using runtime env is the neatest one (at least until `caddy-l4` is configurable via Caddyfile, or there’s some nicer way to load both Caddyfile and JSON config without merging them first).

**Comment by ratorx** [comment]
Also, as a followup, I use `{env.*}` in my `basic_auth` block in the Caddyfile, and it works fine there (i.e. the evaluation of environment at runtime works there). I guess the `http` app is special cased so it can resolve env at runtime (at least inside routes), but the TLS app isn’t?

**Comment by francislavoie** [comment]
Every module needs to specifically invoke replacement for each config value it to work there. It doesn't work everywhere. It depends. We try to make it work wherever we can, some places we may just have missed, and others just can't work for other reasons.

**Comment by mholt** [comment]
@ratorx Can you propose a patch that does what you want? We can look at it in more detail then.

**Comment by ratorx** [comment]
This Caddyfile:
```
{
	acme_ca https://included.just.so.tls.automation.policy.is.generated
}

https://{env.EXAMPLE} {
	respond 200
}
```

generates this JSON if the subject validation error was simply ignored:
```
{
	"apps": {
		"http": {
			"servers": {
				"srv0": {
					"listen": [
						":443"
					],
					"routes": [
						{
							"match": [
								{
									"host": [
										"{env.EXAMPLE}"
									]
								}
							],
							"handle": [
								{
									"handler": "subroute",
									"routes": [
										{
											"handle": [
												{
													"handler": "static_response",
													"status_code": 200
												}
											]
										}
									]
								}
							],
							"terminal": true
						}
					]
				}
			}
		},
		"tls": {
			"automation": {
				"policies": [
					{
						"subjects": [
							"{env.EXAMPLE}"
						],
						"issuers": [
							{
								"ca": "https://included.just.so.tls.automation.policy.is.generated",
								"module": "acme"
							}
						]
					}
				]
			}
		}
	}
}
```

I didn't change any code to get this, I just took the JSON generated when using this Caddyfile:
```
{
	acme_ca https://included.just.so.tls.automation.policy.is.generated
}

https://uniquestring {
	respond 200
}
```

and replaced all `uniquestring` in the generated JSON with `{env.EXAMPLE}`.

The JSON produced by this process was happily accepted and worked with `caddy run`. It correctly tried to get TLS certificate for the value of the environment value rather than for the string `{env.EXAMPLE}`. With a slightly different TLS configuration (because the ACME process will always fail when using the acme_ca in the example), `caddy run` also resolved the top-level route matcher when it was `{env.EXAMPLE}` and listened on the correct address.

So at least in this case, the validation logic when translating Caddyfile into JSON incorrectly rejects JSON config that would have worked. There are two ways I can think of to fix this:

* If the validation logic is necessary, then I can add `{env.*}` expressions to be considered valid subjects.
* Alternatively, the validation logic could be removed.

Which would you prefer?

**Comment by ratorx** [comment]
Sorry, I don't think I have enough knowledge of the codebase to make an actual PR. I had a quick look though and have some ideas of the changes that would be needed (although this is bound to be incomplete):
* Use `caddy.Replacer` inside `AutomationPolicy.Provision` and replace the names of the automated certificates with it (to resolve `{env.*}` and other patterns).
* Either rip out the certificate name validation stuff in Caddyfile to JSON code or special case it to allow patterns.

I guess the tricky bit is handling the current split between internal and external certificates, because there's no way to know without looking at the environment. However, I noticed some stuff inside `TLS.Provision` which seems to be doing the same checks there. Possibly the internal/external split could be omitted and the split could be handled there?

Having thought this through, this seems like a fairly large amount of work for fairly minimal gain (if AutomationPolicy is special cased). I don't think I have the patience to do it, so I definitely understand if you reject this feature request.

However, perhaps if there is enough demand for using patterns in other places, a layer could be introduced to apply the global replacer before any of the config is parsed?

**Comment by mholt** [comment]
Yeah... so this is tricky because the Caddyfile adapter doesn't know how to generate proper JSON if it doesn't know what the site address is, because the properties of the site address dictate how to best create an intuitively correct JSON config.

I think any sort of pre-conditioning or pre-generation should be done before it reaches Caddy, or leave it up to runtime in the actual JSON config.
